File: prototype/wasp/packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scuint64.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

export const ScUint64Length = 8;

import {panic} from '../sandbox';
import {uintFromString, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function uint64Decode(dec: WasmDecoder): u64 {
    return uint64FromBytes(dec.fixedBytes(ScUint64Length));
}

export function uint64Encode(enc: WasmEncoder, value: u64): void {
    enc.fixedBytes(uint64ToBytes(value), ScUint64Length);
}

export function uint64FromBytes(buf: Uint8Array): u64 {
    if (buf.length == 0) {
        return 0;
    }
    if (buf.length != ScUint64Length) {
        panic('invalid Uint64 length');
    }
    let ret: u64 = buf[7];
    ret = (ret << 8) | buf[6];
    ret = (ret << 8) | buf[5];
    ret = (ret << 8) | buf[4];
    ret = (ret << 8) | buf[3];
    ret = (ret << 8) | buf[2];
    ret = (ret << 8) | buf[1];
    return (ret << 8) | buf[0];
}

export function uint64ToBytes(value: u64): Uint8Array {
    const buf = new Uint8Array(ScUint64Length);
    buf[0] = value as u8;
    buf[1] = (value >> 8) as u8;
    buf[2] = (value >> 16) as u8;
    buf[3] = (value >> 24) as u8;
    buf[4] = (value >> 32) as u8;
    buf[5] = (value >> 40) as u8;
    buf[6] = (value >> 48) as u8;
    buf[7] = (value >> 56) as u8;
    return buf;
}

export function uint64FromString(value: string): u64 {
    return uintFromString(value, 64);
}

export function uint64ToString(value: u64): string {
    return value.toString();
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableUint64 {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return uint64ToString(this.value());
    }

    value(): u64 {
        return uint64FromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableUint64 extends ScImmutableUint64 {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: u64): void {
        this.proxy.set(uint64ToBytes(value));
    }
}

File: prototype/wasp/packages/wasmvm/wasmlib/src/wasmtypes/scuint64.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::*;

pub const SC_UINT64_LENGTH: usize = 8;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn uint64_decode(dec: &mut WasmDecoder) -> u64 {
    uint64_from_bytes(&dec.fixed_bytes(SC_UINT64_LENGTH))
}

pub fn uint64_encode(enc: &mut WasmEncoder, value: u64) {
    enc.fixed_bytes(&uint64_to_bytes(value), SC_UINT64_LENGTH);
}

pub fn uint64_from_bytes(buf: &[u8]) -> u64 {
    let len = buf.len();
    if len == 0 {
        return 0;
    }
    if len != SC_UINT64_LENGTH {
        panic("invalid Uint64 length");
    }
    u64::from_le_bytes(buf.try_into().expect("WTF?"))
}

pub fn uint64_to_bytes(value: u64) -> Vec<u8> {
    value.to_le_bytes().to_vec()
}

pub fn uint64_from_string(value: &str) -> u64 {
    value.parse::<u64>().unwrap()
}

pub fn uint64_to_string(value: u64) -> String {
    value.to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableUint64 {
    proxy: Proxy,
}

impl ScImmutableUint64 {
    pub fn new(proxy: Proxy) -> ScImmutableUint64 {
        ScImmutableUint64 { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        uint64_to_string(self.value())
    }

    pub fn value(&self) -> u64 {
        uint64_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable u64 in host container
pub struct ScMutableUint64 {
    proxy: Proxy,
}

impl ScMutableUint64 {
    pub fn new(proxy: Proxy) -> ScMutableUint64 {
        ScMutableUint64 { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: u64) {
        self.proxy.set(&uint64_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        uint64_to_string(self.value())
    }

    pub fn value(&self) -> u64 {
        uint64_from_bytes(&self.proxy.get())
    }
}
